lib/elixir/lib/kernel.ex

```diff
@@ -2173,6 +2173,18 @@ defmodule Kernel do
     Enum.reduce(fields, struct, fun)
   end
 
+  defp validate_struct!(%{__struct__: module} = struct, _module, _arity) when is_atom(module) do
+    struct
+  end
+
+  defp validate_struct!(expr, module, arity) do
+    error_message =
+      "expected #{inspect(module)}.__struct__/#{arity} to return a map with a :__struct__ " <>
+        "key that holds the name of the struct (atom), got: #{inspect(expr)}"
```

Comment:
I finally didn't use `Macro.to_string/1` because I think `__struct__/0,1` is not meant to return an AST.

I see on the Erlang side we are using `'Elixir.Macro':to_string`, when I think `'Elixir.Kernel':inspect` should be used, because it might lead to misinterpretations. Here is an example:

```elixir
iex(1)> defmodule Foo do
...(1)> def __struct__, do: {:ok, :one, :two}
...(1)> def __struct__(_), do: {:ok, :one, :two}
...(1)> end
{:module, Foo,
 <<70, 79, 82, 49, 0, 0, 4, 148, 66, 69, 65, 77, 65, 116, 85, 56, 0, 0, 0, 128,
   0, 0, 0, 13, 10, 69, 108, 105, 120, 105, 114, 46, 70, 111, 111, 8, 95, 95,
   105, 110, 102, 111, 95, 95, 7, 99, 111, ...>>, {:__struct__, 1}}
iex(2)> %Foo{}
** (CompileError) iex:2: expected Foo.__struct__/1 to return a map with a :__struct__ key that holds the name of the struct (atom), got: ok
```

I will fix that (and other occurrences) in a different PR 🙂

Reply:
Good catch.